File: tools/ghidra/ExportSptmNvme.java
// Decompile the stripped T8142 SPTM functions that enforce NVMe queue admission.
//@category AuroraSilicon

import java.io.FileWriter;
import java.util.Map;
import java.util.TreeMap;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.symbol.Reference;


public class ExportSptmNvme extends GhidraScript {
    private static final String OUTPUT = "/tmp/sptm-t8142-nvme-decomp.txt";

    private static final long[] SEED_REFS = {
        0xfffffff0270c6aecL, 0xfffffff0270c6b88L,
        0xfffffff0270c6e0cL, 0xfffffff0270c6ea8L,
        0xfffffff0270c7010L, 0xfffffff0270c706cL, 0xfffffff0270c70d4L,
        0xfffffff0270c73f8L, 0xfffffff0270c749cL, 0xfffffff0270c7544L,
    };

    private static final long[] NVME_STRINGS = {
        0xfffffff027008535L, 0xfffffff027008597L,
        0xfffffff027008675L, 0xfffffff027008703L,
        0xfffffff02700878bL, 0xfffffff0270087f4L,
        0xfffffff027008829L, 0xfffffff0270088afL,
    };

    private void addContainingFunction(Map<Long, Function> functions,
                                       FunctionManager manager, Address address) {
        Function function = manager.getFunctionContaining(address);
        if (function != null) {
            functions.put(function.getEntryPoint().getOffset(), function);
        }
    }

    @Override
    public void run() throws Exception {
        FunctionManager manager = currentProgram.getFunctionManager();
        Map<Long, Function> functions = new TreeMap<>();

        for (long seed : SEED_REFS) {
            addContainingFunction(functions, manager, toAddr(seed));
        }

        for (long stringAddress : NVME_STRINGS) {
            for (Reference reference : getReferencesTo(toAddr(stringAddress))) {
                addContainingFunction(functions, manager, reference.getFromAddress());
            }
        }

        DecompInterface decompiler = new DecompInterface();
        decompiler.openProgram(currentProgram);

        StringBuilder output = new StringBuilder();
        output.append("T8142 SPTM NVME HANDLERS\n");
        output.append("program=").append(currentProgram.getName())
              .append(" image_base=").append(currentProgram.getImageBase()).append("\n\n");

        for (Function function : functions.values()) {
            output.append("==============================================================================\n");
            output.append("FUNCTION ").append(function.getName())
                  .append(" entry=").append(function.getEntryPoint())
                  .append(" body=").append(function.getBody()).append("\n");

            DecompileResults result = decompiler.decompileFunction(function, 120, monitor);
            if (result.decompileCompleted() && result.getDecompiledFunction() != null) {
                output.append(result.getDecompiledFunction().getC());
            } else {
                output.append("DECOMPILATION FAILED: ").append(result.getErrorMessage()).append("\n");
            }
            output.append("\n");
        }

        decompiler.dispose();

        try (FileWriter writer = new FileWriter(OUTPUT)) {
            writer.write(output.toString());
        }

        println("Aurora SPTM export: " + functions.size() + " functions -> " + OUTPUT);
    }
}

File: tools/ghidra/ExportSptmNvme.py
# Decompile the stripped T8142 SPTM functions that enforce NVMe queue admission.
#@category AuroraSilicon

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor


OUTPUT = "/tmp/sptm-t8142-nvme-decomp.txt"

# References to SPTM's NVMe assertion/function-name strings, recovered from the
# matching 25G76 IPSW.  Function containment is more reliable than relying on
# stripped symbol names.
SEED_REFS = [
    0xfffffff0270c6aec,  # sptm_nvme_bar_iocq_reg
    0xfffffff0270c6b88,
    0xfffffff0270c6e0c,  # sptm_nvme_bar_iosq_reg
    0xfffffff0270c6ea8,
    0xfffffff0270c7010,  # sptm_nvme_bar_ioqa_reg
    0xfffffff0270c706c,
    0xfffffff0270c70d4,
    0xfffffff0270c73f8,  # sptm_nvme_bar_admin_queue_regs
    0xfffffff0270c749c,
    0xfffffff0270c7544,
]

NVME_STRINGS = [
    0xfffffff027008535,  # nvme_bootstrap
    0xfffffff027008597,  # sptm_nvme_map_pages
    0xfffffff027008675,  # sptm_nvme_unmap_pages
    0xfffffff027008703,  # sptm_nvme_bar_admin_queue_regs
    0xfffffff02700878b,  # sptm_nvme_bar_ioqa_reg
    0xfffffff0270087f4,  # sptm_nvme_bar_iosq_reg
    0xfffffff027008829,  # sptm_nvme_bar_iocq_reg
    0xfffffff0270088af,  # validate_nvme_call_allowed
]


def containing_function(address):
    return currentProgram.getFunctionManager().getFunctionContaining(toAddr(address))


def add_function(functions, function):
    if function is not None:
        functions[function.getEntryPoint().getOffset()] = function


def main():
    functions = {}

    for address in SEED_REFS:
        add_function(functions, containing_function(address))

    # Also consume any references Ghidra recovered to the named assertion
    # strings. This catches handler functions beyond the known seed range.
    for string_address in NVME_STRINGS:
        for reference in getReferencesTo(toAddr(string_address)):
            add_function(functions, containing_function(reference.getFromAddress().getOffset()))

    decompiler = DecompInterface()
    decompiler.openProgram(currentProgram)
    monitor = ConsoleTaskMonitor()

    output = []
    output.append("T8142 SPTM NVME HANDLERS\n")
    output.append("program=%s image_base=%s\n\n" %
                  (currentProgram.getName(), currentProgram.getImageBase()))

    for entry in sorted(functions.keys()):
        function = functions[entry]
        output.append("=" * 78 + "\n")
        output.append("FUNCTION %s entry=%s body=%s\n" %
                      (function.getName(), function.getEntryPoint(), function.getBody()))

        result = decompiler.decompileFunction(function, 120, monitor)
        if result.decompileCompleted() and result.getDecompiledFunction() is not None:
            output.append(result.getDecompiledFunction().getC())
        else:
            output.append("DECOMPILATION FAILED: %s\n" % result.getErrorMessage())
        output.append("\n")

    decompiler.dispose()

    handle = open(OUTPUT, "w")
    try:
        handle.write("".join(output))
    finally:
        handle.close()

    print("Aurora SPTM export: %d functions -> %s" % (len(functions), OUTPUT))


main()
